### PR TITLE
[flutter_tools] verify checksum of downloaded artifacts

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

All of the network requests from google cloud storage include an x-goog-hash header which contains an MD5 checksum. If present, use to validate that the downloaded binary is valid. This will rule out corrupt files as the cause of getting started crashers in the flutter_tool.

https://github.com/flutter/flutter/issues/38980

This does not fully resolve the above issue, because while we can check if the checksum matches what was expected from cloud storage, this A) may not necessarily be present and B) may not match up to what should be uploaded as part of the engine build process.

But when life gives you lemons you hash those lemons using an outdated hashing algorithm.

### CODE REVIEW

**Comment by jonahwilliams** [comment]
(This also doesn't tell us the hash of the contents in the zipfile)

File: packages/flutter_tools/lib/src/cache.dart [hunk 1]
```diff
@@ -1656,15 +1660,69 @@ class ArtifactUpdater {
...
-      file.writeAsBytesSync(chunk, mode: FileMode.append);
+      inputSink?.add(chunk);
+      randomAccessFile.writeFromSync(chunk);
     });
+    randomAccessFile.closeSync();
+    if (inputSink != null) {
+      inputSink.close();
+      final Digest digest = await digests.stream.last;
+      final String rawDigest = base64.encode(digest.bytes);
+      if (rawDigest != md5Hash) {
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/cache.dart:1695]: This looks like we will throw if `md5Hash` is null, is that correct?

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/cache.dart:1695]: If the md5Hash is null the input sink is not created, there is a test case that verifies this

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/cache.dart:1695]: ahh yeah, i see that now.

**Review by christopherfujino — APPROVED** [review]
LGTM, with a nit question: should we add analytics?

**Comment by jonahwilliams** [comment]
For hash failure?

**Review by christopherfujino — APPROVED** [review]
lgtm

**Comment by christopherfujino** [comment]
> For hash failure?

yep

**Comment by jonahwilliams** [comment]
I'd like to spend more time thinking about the right analytics for this. We should also measure download speed, and the exact type of exception and how many retries.

**Pull request merged**